File: backend/app/schemas/test_sessions.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import ModuleType, TestSessionStatus
from app.schemas.attempts import AttemptResponse


class TestSessionCreate(BaseModel):
    test_version_id: UUID


class SessionAttemptSummary(BaseModel):
    attempt_id: UUID
    module: ModuleType
    status: str
    band_score: float | None
    raw_score: int | None
    max_score: int | None
    elapsed_seconds: int | None


class TestSessionResponse(BaseModel):
    session_id: UUID
    test_version_id: UUID
    test_title: str
    version_number: int
    status: TestSessionStatus
    started_at: datetime
    finished_at: datetime | None
    current_module: ModuleType | None
    next_module: ModuleType | None
    current_attempt: AttemptResponse | None
    attempts: list[SessionAttemptSummary] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    overall_band_score: float | None = None


class TestSessionStartResponse(BaseModel):
    session: TestSessionResponse
    current_attempt: AttemptResponse

File: frontend/src/lib/api/test-sessions.ts
import { z } from "zod";
import { attemptResponseSchema } from "./attempts";
import { apiRequest, type ApiRequester } from "./client";

const moduleSchema = z.enum(["LISTENING", "READING", "WRITING"]);
const sessionAttemptSchema = z.object({
  attempt_id: z.string().uuid(), module: moduleSchema, status: z.string(),
  band_score: z.number().nullable(), raw_score: z.number().int().nullable(),
  max_score: z.number().int().nullable(), elapsed_seconds: z.number().int().nullable(),
});
export const testSessionSchema = z.object({
  session_id: z.string().uuid(), test_version_id: z.string().uuid(), test_title: z.string(),
  version_number: z.number().int(), status: z.enum(["IN_PROGRESS", "COMPLETED", "ABANDONED"]),
  started_at: z.string(), finished_at: z.string().nullable(), current_module: moduleSchema.nullable(),
  next_module: moduleSchema.nullable(), current_attempt: attemptResponseSchema.nullable(),
  attempts: z.array(sessionAttemptSchema), warnings: z.array(z.string()), overall_band_score: z.number().nullable(),
});
export type TestSession = z.infer<typeof testSessionSchema>;

export async function startTestSession(testVersionId: string) {
  const result = await apiRequest<unknown>("/test-sessions", { method: "POST", body: JSON.stringify({ test_version_id: testVersionId }) });
  return z.object({ session: testSessionSchema, current_attempt: attemptResponseSchema }).parse(result);
}
export async function getTestSession(sessionId: string, request: ApiRequester = apiRequest): Promise<TestSession> {
  return testSessionSchema.parse(await request<unknown>(`/test-sessions/${sessionId}`));
}
export async function advanceTestSession(sessionId: string): Promise<TestSession> {
  return testSessionSchema.parse(await apiRequest<unknown>(`/test-sessions/${sessionId}/advance`, { method: "POST" }));
}
